src/System.Linq/tests/ToHashSetTests.cs

```diff
@@ -70,5 +77,362 @@ public void ThrowOnNullSource()
+        }
+
+        [Fact]
+        public void WithGroupBy()
+        {
+            var source = new[] { 1, 2, 3, 1, 2, 3 };
+
+            HashSet<IGrouping<int, int>> result = source.GroupBy(x => x).ToHashSet();
+
+            Assert.NotNull(result);
+            Assert.Equal(3, result.Count);
+        }
+
+        [Fact]
+        public void WithGroupByAndSelector()
+        {
+            var source = new[] { 1, 2, 3, 1, 2, 3 };
+
+            HashSet<IGrouping<int, string>> result = source.GroupBy(x => x, x => x.ToString()).ToHashSet();
+
+            Assert.NotNull(result);
+            Assert.Equal(3, result.Count);
+        }
+
+        [Fact]
+        public void WithGroupByAndElementSelector()
+        {
+            var source = new[] { 1, 2, 3, 1, 2, 3 };
+
+            HashSet<IGrouping<int, string>> result = source.GroupBy(x => x, x => x.ToString()).ToHashSet();
+
+            Assert.NotNull(result);
+            Assert.Equal(3, result.Count);
+        }
+
+        [Fact]
+        public void WithGroupByAndElementAndResultSelector()
+        {
+            var source = new[] { 1, 2, 3, 1, 2, 3 };
+
+            HashSet<IEnumerable<int>> result = source.GroupBy(x => x, x => x, (x, group) => group).ToHashSet();
+
+            Assert.NotNull(result);
+            Assert.Equal(3, result.Count);
+        }
+
+        [Fact]
+        public void WithToLookup()
+        {
+            var source = new[] { 1, 2, 3, 1, 2, 3 };
+
+            HashSet<IGrouping<int, int>> result = source.ToLookup(i => i).ToHashSet();
+
+            Assert.NotNull(result);
+            Assert.Equal(3, result.Count);
+        }
+
+        [Fact]
+        public void WithTake_EmptyPartition()
+        {
+            var source = new int[0];
+
+            // Empty Partition
+            HashSet<int> result = source.Take(1).ToHashSet();
+
+            Assert.NotNull(result);
+            Assert.Equal(0, result.Count);
+        }
+
+        [Fact]
+        public void WithTake_ListPartition()
+        {
+            var source = new[] { 1, 2, 3 }.ToList();
+
+            // List Partition
+            HashSet<int> result = source.Take(1).ToHashSet();
+
+            Assert.NotNull(result);
+            Assert.Equal(1, result.Count);
+        }
+
+        [Fact]
+        public void WithTake_EnumerablePartition()
+        {
+            var source = new[] { 1, 2, 3 };
+
+            // Enumerable Partition
+            HashSet<int> result = source.Take(1).ToHashSet();
+
+            Assert.NotNull(result);
+            Assert.Equal(1, result.Count);
+        }
+
+        [Fact]
+        public void WithSkipThenTake_Partition()
+        {
+            var source = new[] { 1, 2, 3 };
+
+            // Ordered Partition
+            HashSet<int> result = source.Skip(1).Take(1).ToHashSet();
+
+            Assert.NotNull(result);
+            Assert.Equal(1, result.Count);
+        }
+
+        [Fact]
+        public void WithRange()
```

Comment:
Did you add a test for the path where `Range.ToHashSet` is called with a custom comparer and we don't preallocate?